audio/softsynth/sid.cpp

```diff
@@ -1422,6 +1438,28 @@ int SID::updateClock(cycle_count& delta_t, short* buf, int n, int interleave) {
 	return s;
 }
 
+int SID::readBuffer(int16 *buffer, const int numSamples) {
+	int samplesLeft = numSamples;
+
+	while (samplesLeft > 0) {
+		// update SID status after each frame
+		if (_cpuCyclesLeft <= 0) {
+			if (_callback && _callback->isValid())
+				(*_callback)();
```

Comment:
I think you can move the readBuffer implementation to generateSamples as-is. The only problem is the callback handling here; this is already done in Chip::readBuffer. If you remove it in generateSamples it works fine.